### PR TITLE
feat: cache HTTP requests done to the registry

### PR DESCRIPTION
**Pull request by zkochan**
Ref #59

- [x] implement `pnpm cache clean`
- [x] just continue if there was an error during reading from cache
- [x] lock the cache folder to prevent usage of it by several processes
- [x] add tests for checking if cache TTL works

### CODE REVIEW

**Comment by zkochan** [comment]
I expected a bigger performance boost but still. Tests are passing around 20% faster locally

File: docs/api.md [hunk 1]
```diff
@@ -14,6 +14,7 @@ Install packages.
 * `options.global` - *Boolean* - the packages will be installed globally rather than locally.
 * `options.cwd` - *String* - the directory in which the installation will be performed. By default the `process.cwd()` value is used.
 * `options.quiet` - *Boolean* - `false` by default. No output to the console.
+* `options.cacheTTL` - *Number* - `3.600` seconds  (1 hour) by default. The time during which HTTP requests are cached (except the ones that request tarballs).
```
Comment (iamstarkov) [inline docs/api.md]: maybe one day?

File: src/network/got.ts [hunk 2]
```diff
@@ -23,38 +27,49 @@ export type NpmRegistryClient = {
...
 }
 
-export default (client: NpmRegistryClient): Got => {
-  const cache = {}
+export default (client: NpmRegistryClient, opts: {cachePath: string, cacheTTL: number}): Got => {
+  const diskCache = cacheManager.caching({
+    store: fsStore,
+    options: {
+      ttl: opts.cacheTTL,
+      maxsize: 1000 * 1000 * 1000, // max size in bytes on disk
```
Comment (iamstarkov) [inline src/network/got.ts]: 1Gb, right?

**Comment by iamstarkov** [comment]
dont we need some `$ pnpm cache clean` to fix cache in case of fuck up

File: docs/api.md (hunk 1 above)
Comment (zkochan) [reply docs/api.md]: good idea

**Comment by iamstarkov** [comment]
btw, what is the flow of install process has been interrupted either by user, or by any kind of error?

**Comment by iamstarkov** [comment]
> I expected a bigger performance boost but still. Tests are passing around 20% faster locally

as far as i understood from tests, you effectively set cache to 0 in tests, dont you?

File: docs/api.md (hunk 1 above)
Comment (iamstarkov) [reply docs/api.md]: my reasoning is that, if you run pnpm install few times every day, 1hr cache will not help that much

File: src/network/got.ts (hunk 2 above)
Comment (zkochan) [reply src/network/got.ts]: probably less because that would be 1024^3. Probably shouldn't be so big

**Comment by zkochan** [comment]
> as far as i understood from tests, you effectively set cache to 0 in tests, dont you?

I set 0 only in the test that tests update

File: docs/api.md (hunk 1 above)
Comment (iamstarkov) [reply docs/api.md]: on the other side, 1day cache will help pnpm to cache deps for a whole day and user to enjoy fast install process

**Comment by zkochan** [comment]
> dont we need some $ pnpm cache clean to fix cache in case of fuck up

totally. Lets add it

File: src/network/got.ts (hunk 2 above)
Comment (iamstarkov) [reply src/network/got.ts]: lets add 1Gb to a comment as you did with timeout

**Comment by zkochan** [comment]
> btw, what is the flow of install process has been interrupted either by user, or by any kind of error?

you think the cache can get corrupted?

**Comment by zkochan** [comment]
we probably also have to lock the cache folder.

**Comment by iamstarkov** [comment]
> you think the cache can get corrupted?

yes

**Comment by iamstarkov** [comment]
> we probably also have to lock the cache folder.

why? one can run `pnpm i` in two different projects at the same time

**Comment by zkochan** [comment]
yes, but won't the two installs try to write the same cache files?

**Comment by zkochan** [comment]
Regarding the corrupted cache files... I'll add a try/catch and will make the request if the cache reading fails

**Comment by iamstarkov** [comment]
> Regarding the corrupted cache files... I'll add a try/catch and will make the request if reading from cache fails

thats good, but i was thinking about another negative case. take a look yarnpkg/yarn#627

**Comment by iamstarkov** [comment]
> yes, but won't the two installs try to write the same cache files?

oh, do you want to prevent from writing to the same `cache/pkg` folder in particular or to a whole `cache/`

**Comment by zkochan** [comment]
> thats good, but i was thinking about another negative case. take a look yarnpkg/yarn#627

we don't have that issue.

1. we use npm-registry-clinet which handles retries on errors
2. we check the shasum of the tarball after download finished and if they don't match, we throw an error

**Comment by zkochan** [comment]
> oh, do you want to prevent from writing to the same cache/pkg folder in particular or to a whole cache/ folder?

I was thinking about the whole cache/ folder

**Comment by iamstarkov** [comment]
> we use npm-registry-clinet which handles retries on errors

does it? im mostly interested in `content-length` verifying

**Comment by iamstarkov** [comment]
> we check the shasum of the tarball after download finished and if they don't match, we throw an error

awesome

**Comment by zkochan** [comment]
OK, I am not sure about content-length, because that one is in the stream. But we can easily add that check as we already count the size [here](https://github.com/rstacruz/pnpm/blob/master/src/resolve/fetch.ts#L63-L69)

**Comment by zkochan** [comment]
I created a separate issue for the content length checking #423

**Comment by zkochan** [comment]
OK, now I think it is ready.

**Comment by zkochan** [comment]
I merge it to work on the shared storage by default on Node > 6. If you notice something critical, please comment

**Comment by iamstarkov** [comment]
noticed nothing, good job!

**Pull request merged**